compiler/tests-compiler/recfun.ml

```diff
@@ -0,0 +1,110 @@
+(* Js_of_ocaml compiler
+ * http://www.ocsigen.org/js_of_ocaml/
+ * Copyright (C) 2020 Hugo Heuzard
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, with linking exception;
+ * either version 2.1 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
+ *)
+
+open Util
+
+module M = struct
+  type state = int
+
+  type hash_value = int
+
+  let get_hash_value x = x
+
+  let create () = 42
+
+  let fold_int state i = state + i
+
+  let hash_fold_string state s = state + String.length s
+
+  let hash_fold_list fold state l = List.fold_left fold state l
+
+  type t =
+    | Atom of string
+    | List of t list
+
+  let rec (hash_fold_t : state -> t -> state) =
+    (fun hsv arg ->
+       match arg with
+       | Atom _a0 ->
+           let hsv = fold_int hsv 0 in
+           let hsv = hsv in
+           hash_fold_string hsv _a0
+       | List _a0 ->
+           let hsv = fold_int hsv 1 in
+           let hsv = hsv in
+           hash_fold_list hash_fold_t hsv _a0
+      : state -> t -> state)
+
+  and (hash : t -> hash_value) =
+    let func arg =
+      get_hash_value
+        (let hsv = create () in
+         hash_fold_t hsv arg)
+    in
+    fun x -> func x
+end
+
+let%expect_test _ =
+  let program =
+    compile_and_parse
+      {|
+type state = int
+type hash_value = int
+let get_hash_value x = x
+let create () = 42
+let fold_int state i = state + i
+let hash_fold_string state s = state + String.length s
+let hash_fold_list fold state l = List.fold_left fold state l
+
+let myfun x =
+  let module M = struct
+    type t =
+      | Atom of string
+      | List of t list
+
+    let rec (hash_fold_t : state -> t -> state) =
+      (fun hsv ->
+         fun arg ->
+           match arg with
+           | Atom _a0 ->
+             let hsv = fold_int hsv 0 in
+             let hsv = hsv in hash_fold_string hsv _a0
+           | List _a0 ->
+             let hsv = fold_int hsv 1 in
+             let hsv = hsv in hash_fold_list hash_fold_t hsv _a0 : state -> t -> state)
+    and (hash : t -> hash_value) =
```

Comment:
I'm not seeing this function in the output.  Did it get inlined?

Reply:
Yes

Reply:
Could you add a test with functions where none of them get inlined?

Reply:
Note that the inlining was enabled by the last commit. Probably better the review per commits

Reply:
I've tune the test to prevent inlining